olive/optimization/tuning_process.py

```diff
@@ -201,148 +201,177 @@ def create_inference_session(model_path, test_params=None):
         main_process = Process(name="main_process", target=get_throughput,
                                args=(optimization_config, test_params, test_result))
+
+        process_list = []
+        concurrency_num = test_params.get('concurrency_num', 1)
+
+        num_of_background = concurrency_num - 1
+        if num_of_background > 0:
+            synchronizer = Barrier(num_of_background + 1)
+            for i in range(0, num_of_background):
+                p = Process(name="{}_{}".format(SUB_PROCESS_NAME_PREFIX, i), target=get_throughput,
+                            args=(optimization_config, test_params, None, synchronizer))
+                p.start()
+                logger.info("PID [{}] started".format(p.pid))
+                process_list.append(p)
+            synchronizer.wait()
+
+        # execute main func, we only collect benchmark from this main func
+        main_process.start()
+        main_process.join()
+
+        # once the main func finished, stop child process
+        for p in process_list:
+            if p.is_alive():
+                p.terminate()
+                logger.info("PID {} killed".format(p.pid))
+        return dict(test_result)
     else:
-        main_process = Process(name="main_process", target=get_latency,
-                               args=(optimization_config, test_params, test_result))
+        test_result = benchmark_latency(optimization_config, test_params)
+        return test_result
+
 
-    process_list = []
-    concurrency_num = test_params.get('concurrency_num', 1)
+def benchmark_latency(optimization_config, test_params):
+    # Session name will be "pretuning" if there are no parameters
+    session_name = generate_test_name(test_params)
 
-    num_of_background = concurrency_num - 1
-    if num_of_background > 0:
-        synchronizer = Barrier(num_of_background + 1)
-        for i in range(0, num_of_background):
-            p = Process(name="{}_{}".format(SUB_PROCESS_NAME_PREFIX, i), target=concurrent_inference,
-                        args=(synchronizer, optimization_config, test_params))
+    manager = Manager()
+    latencies = manager.list()
+
+    # Do a warm up in the main process to test for an early out
+    should_perform_test_run = True
+    if test_params and not optimization_config.run_all:
+        # Number of tests to determine early out criteria
+        EARLYOUT_TEST_COUNT = 20
+        # Skip test if latencies are X times more than pretuning
+        EARLYOUT_LATENCY_FACTOR = 5
+
+        earlyout_optimization_config = copy(optimization_config)
+        earlyout_optimization_config.test_num = EARLYOUT_TEST_COUNT
+        earlyout_latencies = []
+        get_latency(earlyout_optimization_config, test_params, earlyout_latencies)
+
+        earlyout_latency_ms = sum(earlyout_latencies) / len(earlyout_latencies) * 1000
+        if (earlyout_latency_ms > (earlyout_optimization_config.pretuning_latency_ms * EARLYOUT_LATENCY_FACTOR)):
```

Comment:
What happens if we're tuning for throughput?  In that case, the single-process latency may be poor, but the aggregate throughput across replicas is worth preserving.